File: javascript/data_structures/trees/binary_search_tree.js
/**
 * Binary Search Tree (BST) Data Structure Implementation
 * 
 * A Binary Search Tree is a node-based binary tree data structure that has the following properties:
 * - The left subtree of a node contains only nodes with keys less than the node's key
 * - The right subtree of a node contains only nodes with keys greater than the node's key
 * - Both the left and right subtrees must also be binary search trees
 * 
 * Key characteristics:
 * - Hierarchical structure
 * - Ordered data storage
 * - No duplicate values
 * - Efficient search operations
 * - Dynamic size
 * 
 * Common use cases:
 * - Database indexing
 * - File system organization
 * - Network routing tables
 * - Symbol tables in compilers
 * - Game state management
 * - Decision trees in AI/ML
 * 
 * Operations:
 * - insert: Add a new node with a value
 * - lookup: Find a node with a specific value
 * - remove: Delete a node with a specific value
 * - traverse: Visit all nodes in a specific order
 * 
 * Time Complexity:
 * - Insert: O(log n) average, O(n) worst
 * - Lookup: O(log n) average, O(n) worst
 * - Remove: O(log n) average, O(n) worst
 * - Traverse: O(n)
 * 
 * Space Complexity: O(n) where n is the number of nodes
 * 
 * Note: The performance of BST operations depends on the tree's balance.
 * In worst case (skewed tree), operations can degrade to O(n).
 */

class Node {
  constructor(value) {
    this.value = value;
    this.left = null;
    this.right = null;
  }
}
class BinarySearchTree {
  constructor() {
    this.root = null;
  }

  insert(value) {
    let newNode = new Node(value);
    if (this.root === null) {
      this.root = newNode;
      return console.log(this);
    } else {
      let currentNode = this.root;
      while (true) {
        if (value < currentNode.value) {
          if (!currentNode.left) {
            currentNode.left = newNode;
            return console.log(this);
          }
          currentNode = currentNode.left;
        } else {
          if (!currentNode.right) {
            currentNode.right = newNode;
            return console.log(this);
          }
          currentNode = currentNode.right;
        }
      }
    }
  }

  lookup(value) {
    let currentNode = this.root;
    while (currentNode !== null) {
      if (value < currentNode.value) {
        currentNode = currentNode.left;
      } else if (value > currentNode.value) {
        currentNode = currentNode.right;
      } else {
        return console.log(currentNode);
      }
    }
    return console.log("Not Found");
  }
}

const bst = new BinarySearchTree();
bst.insert(1); //Inserts 1 to the Tree
bst.insert(2); //Inserts 2 to the Tree
bst.insert(0); //Inserts 0 to the Tree
bst.insert(3); //Inserts 3 to the Tree
bst.lookup(2); //Searches for 2 in the Tree

File: python/data_structures/trees/binary_search_tree.py
"""
Binary Search Tree (BST) Implementation

A Binary Search Tree is a node-based binary tree data structure where each node follows the BST property:
For any node N:
- All values in the left subtree of N are less than N’s value
- All values in the right subtree of N are greater than N’s value

Key characteristics:
- Binary tree structure (each node has at most 2 children)
- Ordered structure enables fast lookup, insertion, and deletion
- In-order traversal yields sorted elements

Common use cases:
- Efficient searching and insertion/removal of sorted data
- Implementing associative arrays or sets
- Range queries and sorted data storage
- Underlying structure for maps and sets in many libraries

Operations:
- insert: Adds a new node to the BST in the correct sorted position
- bfs (Breadth-First Search): Traverses the tree level by level
- dfs (Depth-First Search): Traverses the tree in a depthward manner (pre-order here)

Time Complexity:
- Average case: O(log n) for search, insert, delete
- Worst case (unbalanced): O(n)
"""

from collections import deque

class Node:
    def __init__(self, value):
        self.left = None
        self.right = None
        self.value = value

class BST:
    def __init__(self):
        self.root = None

    def insert(self, value):
        node = Node(value)
        if self.root is None:
            self.root = node
            return

        current = self.root
        while True:
            if value > current.value:
                if current.right is not None:
                    current = current.right
                else:
                    current.right = node
                    break
            elif value < current.value:
                if current.left is not None:
                    current = current.left
                else:
                    current.left = node
                    break
            else:
                # Optional: Ignore duplicates
                break

    def bfs(self):
        if self.root is None:
            print([])
            return

        result = []
        queue = deque([self.root])

        while queue:
            node = queue.popleft()
            result.append(node.value)
            if node.left:
                queue.append(node.left)
            if node.right:
                queue.append(node.right)

        print("BFS Traversal:", result)

    def dfs(self):
        if self.root is None:
            print([])
            return

        result = []
        stack = [self.root]

        while stack:
            node = stack.pop()
            result.append(node.value)
            if node.right:
                stack.append(node.right)
            if node.left:
                stack.append(node.left)

        print("DFS (Pre-order) Traversal:", result)
        
    def lookup(self, value):
        current = self.root
        while current is not None:
            if value > current.value:
                current = current.right
            elif value < current.value:
                current = current.left
            else:
                return current  # Node found
        return None  # Node not found


# Example usage
bst = BST()
bst.insert(20)
bst.insert(21)
bst.insert(18)
bst.insert(1)

bst.bfs()  # Output: BFS Traversal: [20, 18, 21, 1]
bst.dfs()  # Output: DFS (Pre-order) Traversal: [20, 18, 1, 21]